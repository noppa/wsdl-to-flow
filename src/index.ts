#!/usr/bin/env node
"use strict";

import { rename, writeFile } from "fs";
import minimist, { ParsedArgs } from "minimist";
import mkdirp from "mkdirp";
import {
  IInterfaceOptions,
  mergeTypedWsdl,
  outputTypedWsdl,
  wsdl2ts
} from "./wsdl-to-ts";

interface IConfigObject {
  outdir: string;
  files: string[];
  tslintDisable: null | string[];
  tslintEnable: null | string[];
}

const opts: IInterfaceOptions = {};
const config: IConfigObject = {
  outdir: "./wsdl",
  files: [],
  tslintDisable: ["max-line-length", "no-empty-interface"],
  tslintEnable: []
};

// eslint-disable-next-line @typescript-eslint/no-unsafe-call
const args: ParsedArgs = minimist(process.argv.slice(2));

if (args.help) {
  // TODO
}

if (args.version) {
  // eslint-disable-next-line
  const pack: { version: number } = require("../package.json");
  console.log("%s %s", "wsdl-to-ts", pack.version);
  process.exit(0);
}

if (args.outdir || args.outDir) {
  config.outdir = String(args.outdir || args.outDir);
}

if (args._) {
  config.files = config.files.concat(args._.map(String));
}

if (config.files.length === 0) {
  console.error("No files given");
  process.exit(1);
}

function mkdirpp(dir: string, mode?: number): Promise<string> {
  return mkdirp(dir, mode || 0o755);
}

Promise.all(config.files.map((a) => wsdl2ts(a, opts)))
  .then(([xs1, ...restXs]) => mergeTypedWsdl(xs1, ...restXs))
  .then(outputTypedWsdl)
  .then((xs: { file: string; data: string[] }[]) => {
    return Promise.all(
      xs.map((x) => {
        console.log("-- %s --", x.file);
        console.log("%s", x.data.join("\n\n"));
        const file = config.outdir + "/" + x.file;
        const dir = file.replace(/\/[^/]+$/, "");
        return mkdirpp(dir).then(() => {
          return new Promise((resolve, reject) => {
            const tsfile = file + ".ts.tmp";
            const fileData: string[] = [];
            if (config.tslintEnable === null) {
              fileData.push("/* tslint:enable */");
            }
            if (config.tslintDisable === null) {
              fileData.push("/* tslint:disable */");
            } else if (config.tslintDisable.length !== 0) {
              fileData.push(
                "/* tslint:disable:" + config.tslintDisable.join(" ") + " */"
              );
            }
            if (config.tslintEnable && config.tslintEnable.length !== 0) {
              fileData.push(
                "/* tslint:enable:" + config.tslintEnable.join(" ") + " */"
              );
            }
            fileData.push(x.data.join("\n\n"));
            fileData.push("");
            writeFile(tsfile, fileData.join("\n"), (err) => {
              if (err) {
                reject(err);
              } else {
                resolve(tsfile);
              }
            });
          });
        });
      })
    );
  })
  .then((files: string[]) =>
    Promise.all(
      files.map((file) => {
        return new Promise((resolve, reject) => {
          const realFile = file.replace(/\.[^.]+$/, "");
          rename(file, realFile, (err) => {
            if (err) {
              reject(err);
            } else {
              resolve(realFile);
            }
          });
        });
      })
    )
  )
  .catch((err) => {
    console.error(err);
    process.exitCode = 3;
  });
