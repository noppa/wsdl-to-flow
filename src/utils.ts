export const hasOwnProperty = (() => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const object$hasOwnProperty = Object.prototype.hasOwnProperty;
  return <T extends Record<string, unknown>>(obj: T, key: keyof T): boolean =>
    object$hasOwnProperty.call(obj, key) as boolean;
})();
