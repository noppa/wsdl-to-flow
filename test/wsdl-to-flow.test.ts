import nock from "nock";
import { join as pathJoin } from "path";
import * as fs from "fs";

const relative = (path: string): string => pathJoin(__dirname, path);

const wsdlUrl = "https://www.w3schools.com/xml/tempconvert.asmx?WSDL";
let hasCache: boolean;
// We'll cache the WSDL request on disk to avoid spamming w3schools
const requestCacheFile = relative("../cache/wsdl-recording-cache.js.lock");
(function initNock() {
  try {
    const nockScript = fs.readFileSync(requestCacheFile, "utf8");
    if (nockScript.trim().length) {
      // eslint-disable-next-line @typescript-eslint/no-implied-eval
      new Function("nock", nockScript)(nock);
      hasCache = true;
      nock.disableNetConnect();
    } else {
      hasCache = false;
    }
  } catch (err) {
    console.warn(err);
    hasCache = false;
  }
  if (!hasCache) {
    nock.recorder.rec({
      // eslint-disable-next-line @typescript-eslint/naming-convention
      dont_print: true
    });
    console.info(
      "No previously recorded cache for WSDL requests, passing them trhough"
    );
  }
})();

import * as soap from "soap";
import { createDefinitionFileContents, wsdl2flow } from "../src/wsdl-to-flow";
import { format } from "prettier";
import flowBinPath from "flow-bin";
import { execFile } from "child_process";
import mkdirp from "mkdirp";

describe("wsdl2flow", () => {
  let client: soap.Client;
  let compiledOutput: undefined | string;
  beforeAll(async () => {
    client = await soap.createClientAsync(wsdlUrl);
  });

  it("should create Flow definitions", async () => {
    const { fileContents } = await wsdl2flow(() =>
      Promise.resolve(client)
    ).then(createDefinitionFileContents);
    expect(fileContents).toMatchSnapshot();
    compiledOutput = fileContents;
  });

  it("should pass Flow check", async () => {
    if (!compiledOutput) {
      throw new Error("Previous test failed");
    }
    const { writeFile } = fs.promises;

    const testJsSource = format(
      `
      // @flow
      /* eslint-disable */
      import type {TempConvertSoap} from 'wsdl-types/TempConvert/TempConvertSoap'
      declare var soapClient: TempConvertSoap;

      soapClient.FahrenheitToCelsius(
        { Fahrenheit: '5'},
        (err, res) => console.log(res.FahrenheitToCelsiusResult)
      )
    `,
      { parser: "babel-flow" }
    );

    await mkdirp(relative("../flow-sadbox/lib"));

    await Promise.all([
      writeFile(relative("../flow-sadbox/soap-client.js"), testJsSource),
      writeFile(
        relative("../flow-sadbox/lib/wsdl-temp-convert.js.flow"),
        compiledOutput
      )
    ]);

    execFile(
      flowBinPath,
      ["check"],
      {
        cwd: relative("../flow-sadbox/")
      },
      (err, stdout) => {
        expect(stdout?.trim()).toBe("Found 0 errors");
      }
    );
  });

  afterAll(() => {
    if (!hasCache) {
      const cache = nock.recorder.play();
      const cacheScript = cache.join(";\n");
      if (cacheScript.trim().length) {
        fs.writeFileSync(requestCacheFile, cacheScript);
      }
      nock.recorder.clear();
    }
  });
});
