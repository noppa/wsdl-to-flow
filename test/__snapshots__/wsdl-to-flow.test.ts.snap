// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wsdl2flow should create Flow definitions 1`] = `
"// @flow
/* eslint-disable */
declare module \\"wsdl-types/TempConvert/TempConvertSoap\\" {
  declare export type FahrenheitToCelsiusInput = {|
    /** s:string(undefined) */
    Fahrenheit: string,
  |};

  declare export type FahrenheitToCelsiusOutput = {|
    /** s:string(undefined) */
    FahrenheitToCelsiusResult: string,
  |};

  declare export type CelsiusToFahrenheitInput = {|
    /** s:string(undefined) */
    Celsius: string,
  |};

  declare export type CelsiusToFahrenheitOutput = {|
    /** s:string(undefined) */
    CelsiusToFahrenheitResult: string,
  |};

  declare export interface TempConvertSoap {
    FahrenheitToCelsius: (
      input: FahrenheitToCelsiusInput,
      cb: (
        err: any | null,
        result: FahrenheitToCelsiusOutput,
        raw: string,
        soapHeader: { [k: string]: any }
      ) => any,
      options?: any,
      extraHeaders?: any
    ) => void;
    FahrenheitToCelsiusAsync: (
      input: FahrenheitToCelsiusInput,
      options?: any,
      extraHeaders?: any
    ) => Promise<FahrenheitToCelsiusOutput>;
    CelsiusToFahrenheit: (
      input: CelsiusToFahrenheitInput,
      cb: (
        err: any | null,
        result: CelsiusToFahrenheitOutput,
        raw: string,
        soapHeader: { [k: string]: any }
      ) => any,
      options?: any,
      extraHeaders?: any
    ) => void;
    CelsiusToFahrenheitAsync: (
      input: CelsiusToFahrenheitInput,
      options?: any,
      extraHeaders?: any
    ) => Promise<CelsiusToFahrenheitOutput>;
  }
}

declare module \\"wsdl-types/TempConvert/TempConvertSoap12\\" {
  declare export type FahrenheitToCelsiusInput = {|
    /** s:string(undefined) */
    Fahrenheit: string,
  |};

  declare export type FahrenheitToCelsiusOutput = {|
    /** s:string(undefined) */
    FahrenheitToCelsiusResult: string,
  |};

  declare export type CelsiusToFahrenheitInput = {|
    /** s:string(undefined) */
    Celsius: string,
  |};

  declare export type CelsiusToFahrenheitOutput = {|
    /** s:string(undefined) */
    CelsiusToFahrenheitResult: string,
  |};

  declare export interface TempConvertSoap12 {
    FahrenheitToCelsius: (
      input: FahrenheitToCelsiusInput,
      cb: (
        err: any | null,
        result: FahrenheitToCelsiusOutput,
        raw: string,
        soapHeader: { [k: string]: any }
      ) => any,
      options?: any,
      extraHeaders?: any
    ) => void;
    FahrenheitToCelsiusAsync: (
      input: FahrenheitToCelsiusInput,
      options?: any,
      extraHeaders?: any
    ) => Promise<FahrenheitToCelsiusOutput>;
    CelsiusToFahrenheit: (
      input: CelsiusToFahrenheitInput,
      cb: (
        err: any | null,
        result: CelsiusToFahrenheitOutput,
        raw: string,
        soapHeader: { [k: string]: any }
      ) => any,
      options?: any,
      extraHeaders?: any
    ) => void;
    CelsiusToFahrenheitAsync: (
      input: CelsiusToFahrenheitInput,
      options?: any,
      extraHeaders?: any
    ) => Promise<CelsiusToFahrenheitOutput>;
  }
}
"
`;
